bin/util.ml

```diff
@@ -30,16 +31,26 @@ let check_path contexts =
 let find_root () =
   let cwd = Sys.getcwd () in
   let rec loop counter ~candidates ~to_cwd dir =
-    let files = Sys.readdir dir |> Array.to_list |> String.Set.of_list in
-    if String.Set.mem files Workspace.filename then
-      cont counter ~candidates:((0, dir, to_cwd) :: candidates) dir ~to_cwd
-    else if Wp.t = Jbuilder && String.Set.exists files ~f:(fun fn ->
-      String.is_prefix fn ~prefix:"jbuild-workspace") then
-      cont counter ~candidates:((1, dir, to_cwd) :: candidates) dir ~to_cwd
-    else if String.Set.mem files Dune_project.filename then
-      cont counter ~candidates:((2, dir, to_cwd) :: candidates) dir ~to_cwd
-    else
-      cont counter ~candidates dir ~to_cwd
+    match Sys.readdir dir with
```

Comment:
are there other places where we use `readdir` where a similar pattern should be used ? IOW, shoud we add a `readdir` version that returns an option?

Reply:
You mean a result? Since we do need to report the concrete error in different error messages.

I think that would be a nice thing to explore. I had a brief look and it didn't seem too easy.